x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/ip/IpPrefix.java

```diff
@@ -0,0 +1,140 @@
+        examples = @Example(file = "ip", tag = "ipPrefix")
+    )
+    public IpPrefix(
+        Source source,
+        @Param(
+            name = "ip",
+            type = { "ip" },
+            description = "IP address of type `ip` (both IPv4 and IPv6 are supported)."
+        ) Expression ipField,
+        @Param(
+            name = "prefixLength",
+            type = { "integer" },
+            description = "Prefix length. If in the range (0, 32), the IP is treated as an IPv4 address. "
+                + "If in the range (32, 128), the IP is treated as an IPv6 address. "
+                + "If the prefix length is out of range, the function returns `null`."
+        ) Expression prefixLengthField
+    ) {
+        super(source, Arrays.asList(ipField, prefixLengthField));
+        this.ipField = ipField;
+        this.prefixLengthField = prefixLengthField;
+    }
+
+    public Expression ipField() {
+        return ipField;
+    }
+
+    public Expression prefixLengthField() {
+        return prefixLengthField;
+    }
+
+    @Override
+    public boolean foldable() {
+        return Expressions.foldable(children());
+    }
+
+    @Override
+    public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        var ipEvaluatorSupplier = toEvaluator.apply(ipField);
+        var prefixLengthEvaluatorSupplier = toEvaluator.apply(prefixLengthField);
+        return new IpPrefixEvaluator.Factory(source(), ipEvaluatorSupplier, prefixLengthEvaluatorSupplier);
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef ip, int prefixLength) {
+        if (ip.length != 16 || prefixLength < 0 || prefixLength > ip.length * 8) {
+            return null;
+        }
+
+        boolean isIpv4 = prefixLength <= 32;
+        // IPv4 is stored in the last 4 bytes
+        int offset = isIpv4 ? 12 : 0;
+        int fullBytes = offset + prefixLength / 8;
+        int remainingBits = prefixLength % 8;
+
+        BytesRef prefix = new BytesRef(new byte[ip.length]);
+
+        // Copy the first full bytes
+        System.arraycopy(ip.bytes, ip.offset, prefix.bytes, 0, fullBytes);
+
+        // Copy the last byte ignoring the trailing bits
+        if (remainingBits > 0) {
+            byte lastByteMask = (byte) (0xFF << (8 - remainingBits));
+            prefix.bytes[fullBytes] = (byte) (ip.bytes[fullBytes] & lastByteMask);
+        }
+
+        return prefix;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.IP;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution ipResolution = isIPAndExact(ipField, sourceText(), FIRST);
+        if (ipResolution.unresolved()) {
+            return ipResolution;
+        }
+
+        return isInteger(prefixLengthField, sourceText(), SECOND);
```

Comment:
There's a helper method `and` that you can use to make this neater:
```suggestion
        TypeResolution ipResolution = isIPAndExact(ipField, sourceText(), FIRST)
        .and(isInteger(prefixLengthField, sourceText(), SECOND));
```

Reply:
I'll start checking the creation date of the functions I use as the template 🤣 
Changing it!